### DISCUSSION TITLE
Custom Tokenizer as pass-through to spacy.Tokenizer.v1 causes F1 to drop during NER training

### DISCUSSION THREAD

**Discussion by oliviercwa** [explosion/spaCy#9209] [Help: Coding & Implementations] [answered]
I have created a customer tokenizer and initially this tokenizer simply allocates the spacy tokenizer v1
```python
self._spacy_tokenizer = spacy.tokenizer.Tokenizer(vocab)
```
The following calls are delegated directly to spacy_tokenizer. Essentially it is just a pass through at this stage

```python
class MyTokenizer
  def __init__(self, vocab, **cfg):
    self._spacy_tokenizer = spacy.tokenizer.Tokenizer(vocab, **cfg)   
  
  def __call__(self, input_data:):
    return self._spacy_tokenizer(input_data)
  
  def from_bytes(self, bytes_data: bytes, exclude=tuple(), **kwargs):
    self._spacy_tokenizer.from_bytes(bytes_data, exclude=exclude, **kwargs)
    return self
```

When I train the spacy NER component, if I use the spacy tokenizer in the configuration I am getting a F1 score of 85
```
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
```

The moment I replace with mine, the F1 falls down to 50
```
tokenizer = {"@tokenizers":"my-tokenizer-v1"}
```

Additional notes and observations:
- Nothing else is changing in the configuration (which uses en_core_web_md model). This is the only change.
- ```from_bytes``` on MyTokenizer is never called which tells me the rules are not loaded properly. If I change the tokenizer factory in the config en_core_web_md folder to tokenizer = {"@tokenizers":"my-tokenizer-v1"}, then ```from_bytes``` is called YET the F1 score during the training remains around 50
- If I intercept the ```@registry.tokenizers("spacy.Tokenizer.v1")``` to create MyTokenizer instead, ```from_bytes``` get called YET the F1 score still remains around 50.
- It seems that declaring 
```
[initialize]
vectors = "en_core_web_md"
```
loads the config associated with it (en_core_web_md\config.cfg) including the tokenizer declared in the config

What am I missing ? How do I get the same training results with a custom tokenizer ?

**Comment by adrianeboyd** [comment, marked as answer]
You don't show how the custom tokenizer is registered so it's hard to know for sure, but I think you can probably see the difference fairly easily if you inspect the output docs, which I would guess just have whitespace tokenization, or you can compare the tokenizer settings, which are probably all `None` rather than regex methods in the custom tokenizer.

Constructing a tokenizer with just `Tokenizer(vocab)` does not include any of the tokenizer settings that would be loaded from the language defaults with `spacy.Tokenizer.v1`.

Take a look at how `spacy.Tokenizer.v1` loads settings from `nlp.Defaults`:

https://github.com/explosion/spaCy/blob/8bda39f088b1f0943ebd12f5ac425dcc74c4e5f0/spacy/language.py#L69-L92

It shouldn't be related to how vectors are loaded. What you see in the initial `spacy train` eval/output shouldn't be affected by serialization, but if the serialization doesn't work correctly, then it's possible you won't see the same results after reloading the trained model.
